### ISSUE TITLE
Implement topojson reader/writer for vector data

### ISSUE THREAD

**Issue by lossyrob** [geotrellis/geotrellis#1289]

**Comment by adamkozuch** [comment]
Hey, I am interested in that project. Maybe it should be added to this year [ideas list].(https://wiki.eclipse.org/Google_Summer_of_Code_2016_Ideas).

**Comment by lossyrob** [comment]
Hey @adamkozuch, thanks for reaching out. I have been meaning to write the project idea out (and a bunch others) on the wiki, just haven't gotten around to it yet.

Do you have experience with TopoJSON? Perhaps we can do some back and forth here to figure out what a good project description should be?

**Comment by adamkozuch** [comment]
I think that's a great idea. I have some experience with GeoJson and now I learning about TopoJSON on https://github.com/mbostock/topojson/wiki. I assume that I should familiarize myself with geotrellis.vector package in general. What do you think?

**Comment by adamkozuch** [comment]
@lossyrob can you give me some hints? For now, I'm not sure what we wanna achieve. As I understand we have topojson format which that shared lines (arcs) . Now sure if after conversion we still would like to keep that shared lines. For instance, if I read geojson and its equivalent in topojson would converted vectors look the same?

**Comment by echeipesh** [comment]
@adamkozuch You are right, we will require adding `Topology` type to `geotrellis.vector`. At am minimum this would be needed to read TopoJSON objects and preserve the arc definitions, its not obvious to me that the slicing process would produce the same arcs for a set of geometries. 

Further it would be very valuable to actually use the properties of TopoJSON collections to perform the topology preserving re-projection and simplification. So implied in that we'd need to add conversion from a geometry/feature collection to `Topology` and back, extract single geometries by name, and other collection like operations that seem appropriate. I think actually performing the simplification would be a stretch goal for this project but there appears to be pretty good literature on how to go about it, so perhaps its not too unrealistic.

**Comment by adamkozuch** [comment]
@echeipesh so for performing the simplification I would have to implement Visvalingam algorithm?

**Comment by echeipesh** [comment]
Yes, Visvalingam  is what is also referred to in https://bost.ocks.org/mike/simplify/ so that seems like a good plan. That and the reference implementation in JS https://github.com/mbostock/topojson look like excellent resource for this.

**Comment by lossyrob** [comment]
We already have it via JTS.
https://github.com/geotrellis/geotrellis/blob/master/vector/src/main/scala/geotrellis/vector/simplify/SimplifyMethods.scala

So this is more about creating the data type, applying the operations to
that type, transitioning between that type and json/that type and our
vector data types.

Yes, Visvalingam is what is also referred to in
https://bost.ocks.org/mike/simplify/ so that seems like a good plan. That
and the reference implementation in JS https://github.com/mbostock/topojson
look like excellent resource for this.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/geotrellis/geotrellis/issues/1289#issuecomment-197937603>

**Comment by adamkozuch** [comment]
@lossyrob I wonder about  that JTS simplification. Will this work with topologies?

**Comment by lossyrob** [comment]
Hey @adamkozuch,

The proposal looks great! Please submit it ASAP as the deadline is approaching...you can edit it after the fact if we find anything wrong with it, but it looks great at first read through. Can you point me to your GSoC 2015 project that was in Scala?

If I understand correctly, Topojson is a set of arcs with properties, so the JTS simplification will work against those arcs. My first impression is that we won't just be able to extent the current geometry types, but we'll have to talk about that in more detail to figure out what the best approach is.

Cheers,
Rob

**Comment by adamkozuch** [comment]
Hi @lossyrob,
Something came out and I cannot participate in GSOC this year. 
I am very sorry. Still I hope that ideas we shared will be useful for that project.
Cheers,
Adam

**Comment by lossyrob** [comment]
@adamkozuch sorry to hear that, I was looking forward to working with you this summer! But, not doing GSoC doesn't mean not participating in GeoTrellis :) I'm looking forward to merging you `geotrellis-slick` changes after we get this release out the door, and looking forward to any other contributions you'd like to make, be it code or conversation. Thanks for all that you've done so far, and best of luck moving forward.

Cheers,
Rob

**Comment by adamkozuch** [comment]
Thank you for understanding. Of course I still would like to make some contributions to geotrellis!

**Comment by mojodna** [comment]
This would be excellent for creating (more debuggable) tiles for Tangram, à la https://mapzen.com/documentation/vector-tiles/use-service/#available-tile-formats
